### PR TITLE
Modernize UART interrupts

### PR DESCRIPTION
**Pull request by bugadani**
## Thank you for your contribution!

We appreciate the time and effort you've put into this pull request.
To help us review it efficiently, please ensure you've gone through the following checklist:

### Submission Checklist 📝
- [x] I have updated existing examples or added new ones (if applicable).
- [x] I have used `cargo xtask fmt-packages` command to ensure that all changed code is formatted correctly.
- [x] My changes were added to the [`CHANGELOG.md`](https://github.com/esp-rs/esp-hal/blob/main/esp-hal/CHANGELOG.md) in the **_proper_** section.
- [x] I have added necessary changes to user code to the [Migration Guide](https://github.com/esp-rs/esp-hal/blob/main/esp-hal/MIGRATING-0.21.md).
- [x] My changes are in accordance to the [esp-rs API guidelines](https://github.com/esp-rs/esp-hal/blob/main/documentation/API-GUIDELINES.md)

#### Extra:
- [x] I have read the [CONTRIBUTING.md guide](https://github.com/esp-rs/esp-hal/blob/main/documentation/CONTRIBUTING.md) and followed its instructions.

#### Description

Chipping away at the present state of inconsistency.

### CODE REVIEW

File: esp-hal/src/uart.rs [hunk 1]
```diff
@@ -954,6 +925,22 @@ where
     }
 }
 
+/// List of exposed UART events.
+#[derive(EnumSetType)]
```
Comment (Dominaezzz) [inline esp-hal/src/uart.rs:929]: You don't wanna derive anything else?

**Review by Dominaezzz — APPROVED** [review]
Pretty straight forward change

File: esp-hal/src/uart.rs (hunk 1 above)
Comment (bugadani) [reply esp-hal/src/uart.rs:929]: I just copied SpiInterrupt without thinking, may add derives there as well.

Comment (bugadani) [reply esp-hal/src/uart.rs]: Good guy EnumSetType gives us a lot for a single derive, I always forget that.

**Review by MabezDev — COMMENTED** [review]
LGTM! Thanks!

In the name of making sure we end up with complete consistency, could you add an entry to the API-GUIDELINES to mention how the interrupt listen/unlisten APIs should be for all drivers?

**Review by MabezDev — APPROVED** [review]
Thank you!

File: esp-hal/MIGRATING-0.21.md [hunk 2]
```diff
@@ -118,3 +118,45 @@ The `Spi<'_, SPI, HalfDuplexMode>::read` and `Spi<'_, SPI, HalfDuplexMode>::writ
...
++serial.interupts().contains(UartInterrupt::AtCmd)
+```
+
+You can now listen/unlisten multiple interrupt bits at once:
+
+```diff
+-uart0.listen_at_cmd();
+-uart0.listen_rx_fifo_full();
++uart0.listen(UartInterrupt::AtCmd | UartConterrupt::RxFifoFull);
+```˛
```
Comment (bugadani) [inline esp-hal/MIGRATING-0.21.md:162]: oh ffs what's this smudge...

**Pull request merged**